<a href="https://colab.research.google.com/github/millionermarbun356-png/Pemerograman-dasar/blob/main/pertemuan%2010.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [ ]:
import pandas as pd

customers_df =pd.read_csv("/content/customers (1).csv")
customers_df.head()

customer_id        customer_name       gender  age  \
0            1        Leanna Busson       Female   30   
1            2  Zabrina Harrowsmith  Genderfluid   69   
2            3      Shina Dullaghan   Polygender   59   
3            4        Hewet McVitie     Bigender   67   
4            5       Rubia Ashleigh   Polygender   30   

                     home_address  zip_code               city  \
0  8606 Victoria TerraceSuite 560      5464      Johnstonhaven   
1      8327 Kirlin SummitApt. 461      8223    New Zacharyfort   
2       269 Gemma SummitSuite 109      5661           Aliburgh   
3       743 Bailey GroveSuite 141      1729  South Justinhaven   
4         48 Hyatt ManorSuite 375      4032     Griffithsshire   

                          state    country  
0            Northern Territory  Australia  
1               South Australia  Australia  
2  Australian Capital Territory  Australia  
3                    Queensland  Australia  
4                    Queensland  Australia

In [ ]:
import pandas as pd

orders_df = pd.read_csv("/content/orders (1).csv")
orders_df.head()

order_id  customer_id  payment order_date delivery_date
0         1           64    30811  2021-8-30    2021-09-24
1         2          473    50490   2021-2-3    2021-02-13
2         3          774    46763  2021-10-8    2021-11-03
3         4          433    39782   2021-5-6    2021-05-19
4         5          441    14719  2021-3-23    2021-03-24

In [ ]:
import pandas as pd

products_df =pd.read_csv("/content/products.csv")
products_df.head()

product_id product_type  product_name size colour  price  quantity  \
0           0        Shirt  Oxford Cloth   XS    red    114        66   
1           1        Shirt  Oxford Cloth    S    red    114        53   
2           2        Shirt  Oxford Cloth    M    red    114        54   
3           3        Shirt  Oxford Cloth    L    red    114        69   
4           4        Shirt  Oxford Cloth   XL    red    114        47   

                                    description  
0  A red coloured, XS sized, Oxford Cloth Shirt  
1   A red coloured, S sized, Oxford Cloth Shirt  
2   A red coloured, M sized, Oxford Cloth Shirt  
3   A red coloured, L sized, Oxford Cloth Shirt  
4  A red coloured, XL sized, Oxford Cloth Shirt

In [ ]:
import pandas as pd

sales_df =pd.read_csv("/content/sales.csv")
sales_df.head()

sales_id  order_id  product_id  price_per_unit  quantity  total_price
0         0         1         218             106         2        212.0
1         1         1         481             118         1        118.0
2         2         1           2              96         3        288.0
3         3         1        1002             106         2        212.0
4         4         1         691             113         3        339.0

In [ ]:
customers_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 6 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   sales_id        5000 non-null   int64  
 1   order_id        5000 non-null   int64  
 2   product_id      5000 non-null   int64  
 3   price_per_unit  5000 non-null   int64  
 4   quantity        5000 non-null   int64  
 5   total_price     4981 non-null   float64
dtypes: float64(1), int64(5)
memory usage: 234.5 KB


In [ ]:

customers_df.isna().sum()

sales_id           0
order_id           0
product_id         0
price_per_unit     0
quantity           0
total_price       19
dtype: int64

In [ ]:
print("Jumlah duplikasi: ", customers_df.duplicated().sum())

Jumlah duplikasi:  0


In [ ]:
customers_df.describe()

sales_id     order_id   product_id  price_per_unit    quantity  \
count  5000.000000  5000.000000  5000.000000     5000.000000  5000.00000   
mean   2499.500000   503.038200   634.053200      103.501600     1.99240   
std    1443.520003   285.964418   363.255794        9.195004     0.80751   
min       0.000000     1.000000     1.000000       90.000000     1.00000   
25%    1249.750000   258.000000   323.000000       95.000000     1.00000   
50%    2499.500000   504.500000   635.000000      102.000000     2.00000   
75%    3749.250000   749.000000   951.000000      112.000000     3.00000   
max    4999.000000   999.000000  1259.000000      119.000000     3.00000   

       total_price  
count  4981.000000  
mean    206.307368  
std      86.352449  
min      90.000000  
25%     112.000000  
50%     204.000000  
75%     285.000000  
max     357.000000

In [ ]:
orders_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 5 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   order_id       1000 non-null   int64 
 1   customer_id    1000 non-null   int64 
 2   payment        1000 non-null   int64 
 3   order_date     1000 non-null   object
 4   delivery_date  1000 non-null   object
dtypes: int64(3), object(2)
memory usage: 39.2+ KB


In [ ]:
print("Jumlah duplikasi: ",orders_df.duplicated().sum())

Jumlah duplikasi:  0


In [ ]:
orders_df.describe()

order_id  customer_id       payment
count  1000.000000  1000.000000   1000.000000
mean    500.500000   506.640000  33972.936000
std     288.819436   277.115502  14451.609047
min       1.000000     1.000000  10043.000000
25%     250.750000   275.250000  21329.250000
50%     500.500000   515.000000  33697.500000
75%     750.250000   737.250000  46249.000000
max    1000.000000  1000.000000  59910.000000

In [ ]:
products_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1266 entries, 0 to 1265
Data columns (total 8 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   product_id    1266 non-null   int64 
 1   product_type  1266 non-null   object
 2   product_name  1266 non-null   object
 3   size          1266 non-null   object
 4   colour        1266 non-null   object
 5   price         1266 non-null   int64 
 6   quantity      1266 non-null   int64 
 7   description   1266 non-null   object
dtypes: int64(3), object(5)
memory usage: 79.3+ KB


In [ ]:
print("Jumlah duplikasi: ", products_df.duplicated().sum())

Jumlah duplikasi:  6


In [ ]:
products_df.describe()

product_id        price     quantity
count  1266.000000  1266.000000  1266.000000
mean    627.926540   105.812006    60.138231
std     363.971586     9.715611    11.682791
min       0.000000    90.000000    40.000000
25%     313.250000    95.250000    50.000000
50%     626.500000   109.000000    60.000000
75%     942.750000   114.000000    70.000000
max    1259.000000   119.000000    80.000000

In [ ]:
sales_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 6 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   sales_id        5000 non-null   int64  
 1   order_id        5000 non-null   int64  
 2   product_id      5000 non-null   int64  
 3   price_per_unit  5000 non-null   int64  
 4   quantity        5000 non-null   int64  
 5   total_price     4981 non-null   float64
dtypes: float64(1), int64(5)
memory usage: 234.5 KB


In [ ]:
sales_df.isna().sum()

sales_id           0
order_id           0
product_id         0
price_per_unit     0
quantity           0
total_price       19
dtype: int64

In [ ]:
print("Jumlah duplikasi: ", sales_df.duplicated().sum())

Jumlah duplikasi:  0


In [ ]:
sales_df.describe()

sales_id     order_id   product_id  price_per_unit    quantity  \
count  5000.000000  5000.000000  5000.000000     5000.000000  5000.00000   
mean   2499.500000   503.038200   634.053200      103.501600     1.99240   
std    1443.520003   285.964418   363.255794        9.195004     0.80751   
min       0.000000     1.000000     1.000000       90.000000     1.00000   
25%    1249.750000   258.000000   323.000000       95.000000     1.00000   
50%    2499.500000   504.500000   635.000000      102.000000     2.00000   
75%    3749.250000   749.000000   951.000000      112.000000     3.00000   
max    4999.000000   999.000000  1259.000000      119.000000     3.00000   

       total_price  
count  4981.000000  
mean    206.307368  
std      86.352449  
min      90.000000  
25%     112.000000  
50%     204.000000  
75%     285.000000  
max     357.000000

In [ ]:
customers_df.drop_duplicates(inplace=True)

In [ ]:
print("Jumlah duplikasi: ", customers_df.duplicated().sum())

Jumlah duplikasi:  0


In [ ]:
customers_df[customers_df.gender.isna()]

Empty DataFrame
Columns: [customer_id, customer_name, gender, age, home_address, zip_code, city, state, country]
Index: []

In [ ]:
customers_df.gender.value_counts()


gender
Male           143
Non-binary     131
Polygender     128
Genderqueer    127
Genderfluid    122
Bigender       120
Female         115
Agender        114
Name: count, dtype: int64

In [ ]:
customers_df.isna().sum()


customer_id      0
customer_name    0
gender           0
age              0
home_address     0
zip_code         0
city             0
state            0
country          0
dtype: int64

In [ ]:
customers_df[customers_df.age == customers_df.age.max()]

customer_id        customer_name  gender  age  \
34            35     Adamo Saddington     1.0   80   
38            39      Legra Neaverson     NaN   80   
142          143      Hewitt Hanscome     NaN   80   
154          155          Myrle Kelby     1.0   80   
170          171  Mariellen Medendorp     NaN   80   
174          175        Patrica Penny     NaN   80   
181          182        Ricardo James     1.0   80   
235          236       Anders Couvert     NaN   80   
423          424     Katherine Sewter     NaN   80   
437          438         Atlante Cure     NaN   80   
494          495         Darci Barten     1.0   80   
523          524      Beverlie Bansal     NaN   80   
596          597           Hi Simenel     NaN   80   
701          702          Magda Dring     NaN   80   
731          732         Gorden Seago     NaN   80   

                          home_address  zip_code                   city  \
34   3168 Bartoletti CrescentSuite 878       593             Port Lucas   
38     7440 Cameron Estate DrSuite 628      4622  North Victoriachester   
142        87 Hahn Station StSuite 943      8254               New Ella   
154           85 Charles MallSuite 424      7841            Port Sophia   
170              16 Boyer WaySuite 018      6226          South Gabriel   
174             57 David MallSuite 596      3129                Haneton   
181           32 Thomas CrestSuite 753      2147             Archiefurt   
235            64 Phillips RunApt. 722      2752            East Sienna   
423            03 Grant CrestSuite 228      5743      Lake Charliemouth   
437          569 Alyssa IslandApt. 766      6893             Hunterland   
494          1798 Schultz MewsApt. 704      9409           Lake Caitlin   
523            0358 Ashley WayApt. 506      3345     Port Sebastianland   
596          78 O'hara CircuitApt. 096      1966       Lake Michaelbury   
701              61 Bosco WaySuite 775       685        Thompsonborough   
731    905 Stephanie BoulevardApt. 217      8114          Port Aaronton   

                  state    country  
34           Queensland  Australia  
38   Northern Territory  Australia  
142            Victoria  Australia  
154     New South Wales  Australia  
170   Western Australia  Australia  
174     South Australia  Australia  
181          Queensland  Australia  
235     New South Wales  Australia  
423   Western Australia  Australia  
437   Western Australia  Australia  
494     New South Wales  Australia  
523          Queensland  Australia  
596     South Australia  Australia  
701   Western Australia  Australia  
731          Queensland  Australia

In [ ]:
customers_df.age.replace(customers_df.age.max(), 70, inplace=True)


/tmp/ipython-input-3463878682.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  customers_df.age.replace(customers_df.age.max(), 70, inplace=True)


In [ ]:
customers_df[customers_df.age == customers_df.age.max()]


customer_id     customer_name  gender  age  \
6              7   Winslow Ewbanck     NaN   76   
85            86  Corinna Maystone     NaN   76   
190          191         Isis Doel     NaN   76   
334          335        Van Winley     1.0   76   
347          348   Thorny Nornable     NaN   76   
378          379        Lilah Mabe     NaN   76   
519          520       Manny Fochs     NaN   76   
541          542      Cass Murrell     NaN   76   
583          584         Page Nise     NaN   76   
678          679   Osgood Percifer     NaN   76   
783          784     Susy De Meyer     1.0   76   
808          809     Winn Siemandl     NaN   76   
896          897      Janeta Berni     1.0   76   

                       home_address  zip_code           city  \
6       92 Hills Station StApt. 683       793      Masonfurt   
85      040 Savannah GroveSuite 967      9337    Gleasonport   
190    540 Schmitt ParkwaySuite 102      5036     Grimesstad   
334       7443 Bayer RidgeSuite 024      5888    Lake George   
347    4070 Nathan CircuitSuite 296      1840       Rubyfort   
378        541 Lucy IslandSuite 157      7863   O'keefemouth   
519      2136 Kemmer IslandApt. 075      1853      Rossshire   
541      870 Waters SquareSuite 227      1416      East Hugo   
583         85 Brown GroveSuite 561      1559   Charlieburgh   
678  7219 Charlotte SummitSuite 507      1971     Cameronton   
783      01 Monahan CircuitApt. 686      8631     Lake Kiara   
808       443 Holly AvenueSuite 631      8563    South Tyler   
896     8325 Molly ParkwaySuite 440      1540  Dickinsonside   

                            state    country  
6                      Queensland  Australia  
85                       Tasmania  Australia  
190            Northern Territory  Australia  
334  Australian Capital Territory  Australia  
347                      Tasmania  Australia  
378                    Queensland  Australia  
519               New South Wales  Australia  
541               New South Wales  Australia  
583                    Queensland  Australia  
678             Western Australia  Australia  
783                      Victoria  Australia  
808                      Tasmania  Australia  
896                    Queensland  Australia

In [ ]:
customers_df.age.replace(customers_df.age.max(), 50, inplace=True)

/tmp/ipython-input-606824494.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  customers_df.age.replace(customers_df.age.max(), 50, inplace=True)


In [ ]:
customers_df[customers_df.age == customers_df.age.max()]

customer_id      customer_name  gender  age  \
7              8       Marlowe Wynn     NaN   75   
15            16  Gabbie Aldwinckle     1.0   75   
24            25      Ernaline Luce     NaN   75   
36            37        Zulema Teml     1.0   75   
63            64   Annabella Devote     NaN   75   
74            75    Katinka Currier     NaN   75   
98            99      Benji Vineall     NaN   75   
116          117     Alexei Hurburt     2.0   75   
141          142   Valerie D'Errico     NaN   75   
224          225        Monty Gowen     NaN   75   
302          303         De Lourens     NaN   75   
331          332     Carla Walework     1.0   75   
337          338        Eve Stutard     2.0   75   
390          391     Reed Churchlow     1.0   75   
408          409    Alphonse Galego     1.0   75   
438          439         Daren Airs     2.0   75   
511          512    Haydon Burnyeat     NaN   75   
539          540       Maude McCaig     NaN   75   
577          578    Riordan Goreway     NaN   75   
603          604    Doralia Nyssens     NaN   75   
627          628    Frayda Simonini     NaN   75   
644          645    Tamarah Costley     2.0   75   
956          957      Felic Marrian     1.0   75   
964          965        Terri Lunny     NaN   75   

                         home_address  zip_code               city  \
7          383 Muller SummitSuite 809      7681            Samside   
15            424 Mason PlaceApt. 181      6438            New Kai   
24         02 Gabriella PlazaApt. 474      9311        Olivershire   
36       3307 Walsh JunctionSuite 233      9751        Violetville   
63          4927 Alice MeadowApt. 960      7787     Sanfordborough   
74           96 Caitlin HillSuite 366      9777          Smythland   
98         468 Shields CircleApt. 480       296  Port Madelineberg   
116           0544 Zoe CourtSuite 153      2398      New Ellieland   
141      607 Parisian AvenueSuite 494      6043        Chelseaview   
224          651 Garden CourtApt. 769      4755        Muellerfurt   
302        334 Olivia MeadowSuite 855      8400        North Paige   
331       011 Hamilton PlaceSuite 215      5702        East George   
337           899 Lynch LaneSuite 349      2652        Monahanberg   
390   4592 Isabella BoulevardApt. 793      7524   North Sophieland   
408      553 Alexander TrailSuite 694       134          New Chase   
438            378 John MallSuite 835      3142        West Thomas   
511           158 Joseph LaneApt. 252      3394        Clarkemouth   
539            297 Ferry LaneApt. 061      8550        East Amelia   
577          256 Andrew CrestApt. 193      8281        East Marcus   
603          17 Connor ParadeApt. 442      4324        Conroymouth   
627  774 Christiansen StreetSuite 261      2197        West Sophie   
644        988 Jacob CircuitSuite 385      1940   Greenfeldershire   
956      5215 Mitchell TrackSuite 778      4220       Samanthaberg   
964          4627 Paige KnollApt. 635      5542     New Callumtown   

                            state    country  
7              Northern Territory  Australia  
15                       Victoria  Australia  
24                       Victoria  Australia  
36              Western Australia  Australia  
63                South Australia  Australia  
74                South Australia  Australia  
98                       Tasmania  Australia  
116                      Tasmania  Australia  
141               South Australia  Australia  
224                    Queensland  Australia  
302               New South Wales  Australia  
331            Northern Territory  Australia  
337               New South Wales  Australia  
390  Australian Capital Territory  Australia  
408  Australian Capital Territory  Australia  
438                      Victoria  Australia  
511  Australian Capital Territory  Australia  
539               South Australia  Australia  
577               New South Wales  Australia  
603  Australi

In [ ]:
customers_df.describe()


customer_id      gender          age     zip_code
count  1000.000000  258.000000  1000.000000  1000.000000
mean    500.500000    1.445736    48.978000  5004.872000
std     288.819436    0.498013    16.585373  2884.497332
min       1.000000    1.000000    20.000000     2.000000
25%     250.750000    1.000000    34.000000  2401.750000
50%     500.500000    1.000000    50.000000  5083.000000
75%     750.250000    2.000000    64.000000  7460.250000
max    1000.000000    2.000000    75.000000  9998.000000

teks Berdasarkan hasil tersebut dapat dilihat bahwa kolom age memiliki nilai maksimum yang cukup masuk akal. Selain itu, jika Ada perhatikan, nilai mean dan standard deviation ikut berubah setelah kita menangani inaccurate value tersebut.

Membersihkan Data orders_df Sekarang kita telah menyelesaikan semua permasalahan yang ada pada customers_df. Selanjutnya, kita akan mengatasi permasalahan pada orders_df. Berdasarkan proses assessing data sebelumnya, diketahui bahwa terdapat kesalahan tipe data untuk kolom order_date & delivery_date.

Untuk mengatasi masalah ini, kita akan mengganti tipe data pada kolom order_date & delivery_date menjadi datetime. Proses ini dapat dilakukan menggunakan function to_datetime() yang disediakan oleh library pandas

In [ ]:
datetime_columns = ["order_date", "delivery_date"]

for column in datetime_columns:
  orders_df[column] = pd.to_datetime(orders_df[column])

Kode di atas akan mengubah tipe data pada kolom order_date & delivery_date menjadi datetime. Untuk memastikan hal ini berjalan sesuai harapan, periksa kembali tipe data tersebut menggunakan method info().

In [ ]:
orders_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 5 columns):
 #   Column         Non-Null Count  Dtype         
---  ------         --------------  -----         
 0   order_id       1000 non-null   int64         
 1   customer_id    1000 non-null   int64         
 2   payment        1000 non-null   int64         
 3   order_date     1000 non-null   datetime64[ns]
 4   delivery_date  1000 non-null   datetime64[ns]
dtypes: datetime64[ns](2), int64(3)
memory usage: 39.2 KB


ika semua tahapan berjalan sesuai harapanm, Maka tipe data kedua kolom sudah berganti menjadi datetime

Membersihkan Data product_df

Data berikutnya yang akan bersihkan ialah product_df. Sesuai hasil assessing data sebelumnya, kita mengetahui bahwa terdapat 6 data yang terduplikasi pada product_df. Untuk mengatasi hal ini, kita perlu membuang data yang sama tersebut menggunakan method drop_duplicates() seperti contoh berikut.

In [ ]:
products_df.drop_duplicates(inplace=True)

Kode di atas akan menghapus semua data yang duplikat. Untuk memastikan kode tersebut berjalan sesuai harapan, jalankanlah kode berikut.

In [ ]:
print("Jumlah duplikasi: ", products_df.duplicated().sum())


Jumlah duplikasi:  0


Jika proses penghapusan data yang duplikat berjalan dengan lancar, kode di atas akan menghasilkan keluaran seperti berikut “Jumlah duplikasi: 0”.

Membersihkan Data sales_df

Data selanjutnya yang perlu bersihkan ialah sales_df. Berdasarkan hasil penilain data sebelumnya, diketahui bahwa terdapat 19 missing value pada kolom total_price. Untuk mengetahui proses penanganan missing value yang paling sesuai, kita perlu melihat terlebih dahulu baris data yang mengandung missing value tersebut.

In [ ]:
sales_df[sales_df.total_price.isna()]

sales_id  order_id  product_id  price_per_unit  quantity  total_price
9            9         2        1196             105         1          NaN
121        121        27        1027              90         3          NaN
278        278        63         360              94         2          NaN
421        421        95        1091             115         1          NaN
489        489       108        1193             105         3          NaN
539        539       117         405             119         2          NaN
636        636       134         653              93         3          NaN
687        687       145        1138             102         1          NaN
854        854       177          64             104         1          NaN
1079      1079       222         908              94         3          NaN
1193      1193       248        1121             102         2          NaN
1313      1313       272         826             117         1          NaN
1548      1548       316         103             118         3          NaN
1688      1688       345         428             107         1          NaN
1775      1775       359         694             113         2          NaN
1902      1902       381        1218             105         3          NaN
2025      2025       408         611             112         3          NaN
2164      2164       436         583             100         3          NaN
2347      2347       476         696             113         2          NaN

Kode di atas akan menampilkan semua baris data yang memiliki missing value pada kolom total_price seperti pada gambar di atas.

Berdasarkan tampilan data tersebut, kita menemukan bahwa nilai total_price merupakan hasil perkalian antara price_per_unit dan quantity.

Kita dapat menggunakan pola ini untuk menangani missing value pada kolom total_price. Berikut merupakan contoh penerapan kode untuk melakukannya.

In [ ]:
sales_df["total_price"] = sales_df["price_per_unit"] * sales_df["quantity"]

Kode di atas akan mengatasi seluruh missing value serta memastikan nilai pada kolom total_price telah sesuai. Untuk memastikan hal ini, Anda dapat memeriksa kembali jumlah missing value pada sales_df menggunakan kode berikut.

In [ ]:
sales_df.isna().sum()

sales_id          0
order_id          0
product_id        0
price_per_unit    0
quantity          0
total_price       0
dtype: int64

Jika proses sebelumnya berjalan lancar, Anda akan menemukan hasil seperti gambar di atas.

Tahapan di atas adalah Proses data wrangling untuk menyiapkan data sebelum dianalisis

Exploratory Data Analysis

Setelah memahami semua teori tentang proses EDA, kini saatnya kita menerapkannya dalam sebuah proyek analisis data sederhana.

Menentukan Pertanyaan Bisnis

Seperti yang telah kita bahas sebelumnya, untuk menentukan sebuah pertanyaan, kita perlu memahami permasalahan bisnis yang dihadapi terlebih dahulu. Nah, pada proyek ini, kita dihadapkan dengan sebuah contoh kasus dari perusahaan online fashion bernama Fashion Collection (FCO).

Sebagai perusahaan yang bergerak di bidang online fashion, Fashion Collection (FCO) perlu mengevaluasi performa penjualan (sales) dan memahami item fashion yang paling banyak dibeli. Selain itu, ia juga perlu lebih memahami pelanggannya (customer) sehingga dapat membuat sebuah strategi campaign yang lebih efisien.

Berdasarkan beberapa masalah tersebut, kita dapat mendefinisikan beberapa pertanyaan bisnis seperti berikut.

1.Bagaimana performa penjualan dan revenue perusahaan dalam beberapa bulan terakhir?

2.Produk apa yang paling banyak dan paling sedikit terjual?
3.Bagaimana demografi pelanggan yang kita miliki?
4.Kapan terakhir pelanggan melakukan transaksi?
5.Seberapa sering seorang pelanggan melakukan pembelian dalam beberapa bulan terakhir?
6.Berapa banyak uang yang dihabiskan pelanggan dalam beberapa bulan terakhir?.

keenam pertanyaan bisnis yang harus kita jawab melalui proses analisis data

jawaban

1.Analisis:

Kelompokkan data sales berdasarkan bulan.

Hitung total penjualan (jumlah transaksi) dan total revenue per bulan.

Gunakan grafik garis/batang untuk melihat tren.

Tujuan:
Mengetahui apakah penjualan meningkat, menurun, atau stabil.

2.Analisis:

Gabungkan data sales dan products.

Hitung total quantity terjual per produk.

Urutkan dari terbesar ke terkecil.

Tujuan:
Mengetahui produk terlaris dan kurang diminati.

3.Analisis:

Gunakan data customers.

Analisis berdasarkan usia, gender, lokasi (kota/negara).

Hitung jumlah pelanggan per kategori.

Tujuan:
Memahami profil mayoritas pelanggan.

4.Analisis:

Ambil tanggal transaksi terakhir per pelanggan dari data orders/sales.

Hitung selisih dengan tanggal hari ini.

Tujuan:
Mengidentifikasi pelanggan aktif dan tidak aktif.

5.Analisis:

Hitung jumlah transaksi per pelanggan dalam periode tertentu.

Kelompokkan pelanggan berdasarkan frekuensi (sering, sedang, jarang).

Tujuan:
Menilai loyalitas pelanggan.

6.Analisis:

Hitung total revenue per pelanggan.

Urutkan dari yang paling besar hingga kecil.

Tujuan:
Menentukan pelanggan dengan nilai tertinggi (high-value customer).



**teksMelakukan Eksplorasi Data tebal**

**Eksplorasi Data customers_df

Pertama, kita akan mengeksplorasi data customers_df terlebih dahulu. Seperti yang kita ketahui, dataset ini memuat berbagai informasi terkait customer, seperti customer_id, customer_name, gender, age, home_address, zip_code, city, state, dan country.

melihat rangkuman parameter statistik dari data customers_df menggunakan method describe() tebal**

In [ ]:
customers_df.describe(include="all")

customer_id     customer_name      gender          age  \
count   1000.000000              1000  258.000000  1000.000000   
unique          NaN              1000         NaN          NaN   
top             NaN  Mandel Fairbanks         NaN          NaN   
freq            NaN                 1         NaN          NaN   
mean     500.500000               NaN    1.445736    48.978000   
std      288.819436               NaN    0.498013    16.585373   
min        1.000000               NaN    1.000000    20.000000   
25%      250.750000               NaN    1.000000    34.000000   
50%      500.500000               NaN    1.000000    50.000000   
75%      750.250000               NaN    2.000000    64.000000   
max     1000.000000               NaN    2.000000    75.000000   

                      home_address     zip_code         city            state  \
count                         1000  1000.000000         1000             1000   
unique                        1000          NaN          961                8   
top     1671 Lauren KnollSuite 945          NaN  East Sophia  South Australia   
freq                             1          NaN            3              139   
mean                           NaN  5004.872000          NaN              NaN   
std                            NaN  2884.497332          NaN              NaN   
min                            NaN     2.000000          NaN              NaN   
25%                            NaN  2401.750000          NaN              NaN   
50%                            NaN  5083.000000          NaN              NaN   
75%                            NaN  7460.250000          NaN              NaN   
max                            NaN  9998.000000          NaN              NaN   

          country  
count        1000  
unique          1  
top     Australia  
freq         1000  
mean          NaN  
std           NaN  
min           NaN  
25%           NaN  
50%           NaN  
75%           NaN  
max           NaN

Berdasarkan rangkuman parameter statistik di atas, kita akan memperoleh informasi jumlah pelanggan sebanyak 1001 orang yang berumur di kisaran antara 20 hingga 80 tahun dengan rata-rata umur sebesar 49.87 tahun dengan standar deviasinya sebesar 17.64 tahun. Informasi ini tentunya dapat memberikan gambaran bahwa segmentasi pelanggan yang kita miliki cukup luas mulai dari remaja hingga lanjut usia.

Sekarang coba kita lihat demografi pelanggan berdasarkan jenis kelamin (gender). Untuk melakukan ini, kita akan menggunakan method groupby() yang diikuti dengan method agg(). Berikut contoh kode untuk melakukannya.

In [ ]:
customers_df.groupby(by="gender").agg({
    "customer_id": "nunique",
    "age": ["max", "min", "mean", "std"]
})

customer_id age                          
           nunique max min       mean        std
gender                                          
1.0            143  75  20  49.937063  17.063765
2.0            115  75  20  48.721739  15.983296

Kode di atas akan menghasilkan sebuah pivot table. Ia memuat informasi jumlah pelanggan (nunique dari customer_id) serta parameter umur yang dikelompokkan berdasarkan jenis kelamin.

Berdasarkan pivot table di atas, dapat diketahui bahwa pelanggan yang kita miliki didominasi oleh jenis kelamin prefer not to say. Di lain sisi, distribusi umur mereka ternyata cukup mirip yaitu berkisar antara 20 hingga 80 tahun.

Selanjutnya, kita coba melihat persebaran jumlah pelanggan berdasarkan kota (city) dan negara bagian (state). Untuk melakukannya kita juga akan menggunakan method groupby(). Selain itu, agar hasilnya lebih mudah untuk dilihat, kita akan mengurutkan nilainya menggunakan method sort_values() secara descending.

In [ ]:
customers_df.groupby(by="city").customer_id.nunique().sort_values(ascending=False)

city
East Sophia     3
New Ava         3
East Aidan      3
Lake Sienna     2
Jademouth       2
               ..
Justinport      1
Kaiberg         1
Kaifort         1
Kaihaven        1
Johnsonhaven    1
Name: customer_id, Length: 961, dtype: int64

In [ ]:
customers_df.groupby(by="state").customer_id.nunique().sort_values(ascending=False)

state
South Australia                 139
Queensland                      134
New South Wales                 132
Northern Territory              125
Western Australia               124
Australian Capital Territory    121
Victoria                        121
Tasmania                        104
Name: customer_id, dtype: int64

Berdasarkan hasil di atas, dapat dilihat bahwa persebaran pelanggan kita cukup merata pada setiap kota dan negara bagian. Pelanggan kita paling banyak berada di kota East Aidan, East Sophia, dan New Ava dengan jumlah masing-masing tiga pelanggan. Selain itu, pelanggan kita paling banyak berasal dari negara bagian South Australia.

Oke, sekarang kita telah memiliki cukup banyak informasi terkait data customers_df. Selanjutnya kita akan melakukan eksplorasi terhadap data orders_df

Eksplorasi Data orders_df

Data kedua yang akan kita eksplor adalah orders_df. Ia mengandung berbagai informasi terkait sebuah order yang terdiri dari order_id, customer_id, order_date, dan delivery_date

Berdasarkan informasi tersebut, kita bisa membuat sebuah kolom baru untuk memuat waktu pengiriman tiap order. Untuk melakukannya, kita perlu menghitung selisih antar delivery_date dan order_date serta menyimpannya sebagai delivery_time. Selanjutnya, kita akan menggunakan method apply() untuk melakukan sebuah operasi terhadap setiap elemen dalam sebuah kolom DataFrame atau Series (bentuk satu dimensi dari DataFrame).

Operasi yang akan kita lakukan ialah menghitung jumlah detik dari delivery_time menggunakan method total_seconds(). selanjutnya diubah ke dalam satuan hari (dibagi 86400) dan diambil bilangan bulatnya

In [ ]:
delivery_time = orders_df["delivery_date"] - orders_df["order_date"]
delivery_time = delivery_time.apply(lambda x: x.total_seconds())
orders_df["delivery_time"] = round(delivery_time/86400)


kode di atas, akan menemukan kolom baru pada DataFrame orders_df

In [ ]:
orders_df.describe(include="all")

order_id  customer_id       payment                  order_date  \
count  1000.000000  1000.000000   1000.000000                        1000   
mean    500.500000   506.640000  33972.936000  2021-05-27 18:38:52.800000   
min       1.000000     1.000000  10043.000000         2021-01-01 00:00:00   
25%     250.750000   275.250000  21329.250000         2021-03-13 18:00:00   
50%     500.500000   515.000000  33697.500000         2021-05-27 12:00:00   
75%     750.250000   737.250000  46249.000000         2021-08-12 06:00:00   
max    1000.000000  1000.000000  59910.000000         2021-10-24 00:00:00   
std     288.819436   277.115502  14451.609047                         NaN   

             delivery_date  delivery_time  
count                 1000    1000.000000  
mean   2021-06-10 20:31:12      14.078000  
min    2021-01-03 00:00:00       1.000000  
25%    2021-03-28 00:00:00       8.000000  
50%    2021-06-11 12:00:00      14.000000  
75%    2021-08-24 06:00:00      21.000000  
max    2021-11-20 00:00:00      27.000000  
std                    NaN       7.707225

Berdasarkan hasil tersebut, dapat diketahui bahwa rata-rata waktu pengiriman sebesar 14 hari dengan nilai maksimum sebesar 27 hari dan nilai minimum sebesar 1 hari.

Eksplorasi Data orders_df dan customers_df

Jika Anda perhatikan, pada data orders_df terdapat kolom yang berisi informasi terkait customer id pelanggan yang pernah melakukan order. Informasi ini bisa kita gunakan untuk mengidentifikasi pelanggan yang belum pernah melakukan order.

Untuk melakukan hal ini, kita bisa membuat sebuah kolom baru bernama “status” pada data customers_df. Kolom tersebut memiliki nilai “Active” untuk pelanggan yang pernah melakukan order setidaknya sekali dan sebaliknya bernilai “Non Active” untuk pelanggan yang belum pernah melakukan order sama sekali

In [ ]:
customer_id_in_orders_df =  orders_df.customer_id.tolist()
customers_df["status"] = customers_df["customer_id"].apply(lambda x: "Active" if x in customer_id_in_orders_df else "Non Active")
customers_df.sample(5)

customer_id      customer_name  gender  age  \
911          912  Chancey Iacovozzo     NaN   60   
941          942     Lacee Denyukin     NaN   45   
756          757       Taffy Quimby     NaN   31   
689          690       Arlie Bowman     NaN   34   
576          577     Hillier Seward     NaN   32   

                         home_address  zip_code               city  \
911           501 Nader RidgeApt. 355      7229          West John   
941     040 Patrick Estate DrApt. 254      5149  North Nicholaston   
756  1033 Botsford Estate DrSuite 236      4773        East Xavier   
689        04 Christian RidgeApt. 329      6697     Port Annabelle   
576         11 Savannah LaneSuite 334      6269         Heaneyview   

                  state    country  status  
911            Tasmania  Australia  Active  
941            Victoria  Australia  Active  
756     New South Wales  Australia  Active  
689  Northern Territory  Australia  Active  
576     South Australia  Australia  Active

Kode di atas akan menghasilkan lima sample data, Bisa di ubah parameter 5 nya sesuai dengan keinginan untuk melihat data yg lebih banyak

Untuk memperoleh informasi terkait jumlah pelanggan yang berstatus “Active” dan “Non Active”, kita bisa menggunakan pivot table

In [ ]:
customers_df.groupby(by="status").customer_id.count()

status
Active        617
Non Active    383
Name: customer_id, dtype: int64

Seperti yang terlihat pada pivot table di atas, terdapat cukup banyak pelanggan yang belum pernah melakukan transaksi sebelumnya. Ini tentunya merupakan data yg buruk karena hampir 30% pelanggan belum pernah melakukan order sebelumnya.

Untuk memperoleh lebih banyak informasi terkait kedua data tersebut, kita perlu menggabungkan keduanya melalui proses join atau merge. Berikut merupakan contoh kode untuk melakukan merge terhadap data orders_df dan customers_df.

In [ ]:
orders_customers_df = pd.merge(
    left=orders_df,
    right=customers_df,
    how="left",
    left_on="customer_id",
    right_on="customer_id"
)
orders_customers_df.head()


order_id  customer_id  payment order_date delivery_date  delivery_time  \
0         1           64    30811 2021-08-30    2021-09-24           25.0   
1         2          473    50490 2021-02-03    2021-02-13           10.0   
2         3          774    46763 2021-10-08    2021-11-03           26.0   
3         4          433    39782 2021-05-06    2021-05-19           13.0   
4         5          441    14719 2021-03-23    2021-03-24            1.0   

         customer_name  gender  age                   home_address  zip_code  \
0     Annabella Devote     NaN   75      4927 Alice MeadowApt. 960      7787   
1          Lori Briars     1.0   61  531 Schmitt BoulevardApt. 010      1744   
2         Ellynn Korba     NaN   34       2096 Wilson MewsApt. 714      8590   
3      Candis Roswarne     NaN   46        5777 Mayer PassApt. 881      9728   
4  Artemas Vasilischev     NaN   53   33 Richards JunctionApt. 478      7650   

                 city              state    country  status  
0      Sanfordborough    South Australia  Australia  Active  
1             Annaton    South Australia  Australia  Active  
2  West Jonathanshire           Tasmania  Australia  Active  
3    West Michaelport           Tasmania  Australia  Active  
4     South Rileyview  Western Australia  Australia  Active

Pada kode di atas, kita melakukan proses merge dengan metode “left join”. metode ini memungkinkan kita untuk mengambil semua nilai dari tabel kiri serta nilai yang bersesuaian dengan tabel kanan.

Sehingga Terdapat banyak sekali informasi yang bisa kita eksplore dari data di atas. Berikut merupakan beberapa hal yang bisa dijelajahi.

Jumlah order berdasarkan kota Kita bisa membuat pivot table untuk memperoleh informasi terkait jumlah order berdasarkan kota dengan kode seperti berikut

In [ ]:
orders_customers_df.groupby(by="city").order_id.nunique().sort_values(ascending=False).reset_index().head(10)

city  order_id
0        Jordanside         6
1           New Ava         6
2  Port Hannahburgh         5
3         Lake Rose         5
4        O'keefeton         5
5          East Max         5
6          Rubyfort         5
7          West Kai         5
8         Sarahstad         4
9        Russelstad         4

berdasarkan hasil tersebut, diketahui bahwa kota Jordanside dan New Ava merupakan dua kota yang memiliki jumlah order terbanyak.

Jumlah order berdasarkan state Sekarang kita akan melihat jumlah order berdasarkan state. Untuk memperoleh informasi terkait hal ini, kita juga bisa menggunakan pivot table. Gunakan kode di bawah ini untuk menerapkannya.

In [ ]:
orders_customers_df.groupby(by="state").order_id.nunique().sort_values(ascending=False)

state
South Australia                 148
Queensland                      139
Western Australia               130
New South Wales                 129
Victoria                        118
Australian Capital Territory    118
Tasmania                        112
Northern Territory              106
Name: order_id, dtype: int64

ode tersebut akan menghasilkan pivot table seperti di bawah ini. Berdasarkan pivot table tersebut, diketahui bahwa South Australia merupakan state yang paling banyak melakukan order.

Jumlah order berdasarkan gender Informasi selanjutnya yang dapat kita eksplorasi adalah jumlah order berdasarkan gender. Untuk melakukan hal ini, tentunya kita perlu membuat sebuah pivot table menggunakan kode berikut.

In [ ]:
orders_customers_df.groupby(by="gender").order_id.nunique().sort_values(ascending=False)


gender
2.0    139
1.0    136
Name: order_id, dtype: int64

Berdasarkan hasil tersebut, dapat dilihat bahwa kebanyakan order dilakukan oleh pelanggan yang berjenis kelamin prefer not to say. Hal ini tentunya sejalan dengan jumlah pelanggan yang kebanyakan dari kelompok gender tersebut.

Jumlah order berdasarkan kelompok usia

Mengeksplorasi jumlah order berdasarkan kelompok usia. Untuk melakukan hal ini, kita perlu mendefinisikan sebuah kolom baru bernama “age_group”. Kolom ini akan membantu kita dalam mengelompokkan pelanggan ke dalam tiga kelompok, yaitu youth, adults, dan seniors. Selanjutnya, buatlah pivot table berdasarkan hal tersebut menggunakan contoh kode di bawah ini.

In [ ]:
orders_customers_df["age_group"] = orders_customers_df.age.apply(lambda x: "Youth" if x <= 24 else ("Seniors" if x > 64 else "Adults"))


In [ ]:
orders_customers_df.groupby(by="age_group").order_id.nunique().sort_values(ascending=False)


age_group
Adults     694
Seniors    213
Youth       93
Name: order_id, dtype: int64

Berdasarkan hasil tersebut diketahui bahwa pelanggan yang paling banyak melakukan order berasal dari kelompok usia Adults.

sejauh ini kita telah menemukan banyak sekali informasi yang tentunya menarik untuk ditelusuri lebih dalam. Namun, sebelum itu mari kita mengeksplorasi dua data yang lain (product_df dan sales_df) terlebih dahulu.

Eksplorasi Data product_df dan sales_df

Pada tahap ini, kita akan mengeksplorasi data product_df dan sales_df. Sebagai permulaan, kita dapat melihat rangkuman parameter statistik dari keduanya menggunakan method describe().

In [ ]:
products_df.describe(include="all")

product_id product_type product_name  size colour        price  \
count   1260.000000         1260         1260  1260   1260  1260.000000   
unique          NaN            3           35     8      7          NaN   
top             NaN        Shirt        Denim    XS    red          NaN   
freq            NaN          420           70   252    180          NaN   
mean     629.500000          NaN          NaN   NaN    NaN   105.805556   
std      363.874979          NaN          NaN   NaN    NaN     9.704423   
min        0.000000          NaN          NaN   NaN    NaN    90.000000   
25%      314.750000          NaN          NaN   NaN    NaN    95.750000   
50%      629.500000          NaN          NaN   NaN    NaN   108.500000   
75%      944.250000          NaN          NaN   NaN    NaN   114.000000   
max     1259.000000          NaN          NaN   NaN    NaN   119.000000   

           quantity                                        description  
count   1260.000000                                               1260  
unique          NaN                                               1260  
top             NaN  A violet coloured, XL sized, Tracksuit Bottoms...  
freq            NaN                                                  1  
mean      60.150000                                                NaN  
std       11.670573                                                NaN  
min       40.000000                                                NaN  
25%       50.000000                                                NaN  
50%       60.000000                                                NaN  
75%       70.000000                                                NaN  
max       80.000000                                                NaN

In [ ]:
sales_df.describe(include="all")

sales_id     order_id   product_id  price_per_unit    quantity  \
count  5000.000000  5000.000000  5000.000000     5000.000000  5000.00000   
mean   2499.500000   503.038200   634.053200      103.501600     1.99240   
std    1443.520003   285.964418   363.255794        9.195004     0.80751   
min       0.000000     1.000000     1.000000       90.000000     1.00000   
25%    1249.750000   258.000000   323.000000       95.000000     1.00000   
50%    2499.500000   504.500000   635.000000      102.000000     2.00000   
75%    3749.250000   749.000000   951.000000      112.000000     3.00000   
max    4999.000000   999.000000  1259.000000      119.000000     3.00000   

       total_price  
count  5000.000000  
mean    206.360000  
std      86.357457  
min      90.000000  
25%     112.000000  
50%     204.000000  
75%     285.000000  
max     357.000000

Berdasarkan hasil tersebut, dapat dilihat bahwa harga barang yang dijual berkisar antara 90 hingga 119 dolar. Selain itu, kita juga memperoleh informasi lain yang tidak kalah menarik yaitu pada setiap transaksi pelanggan paling banyak membeli tiga buah item dalam satu jenis produk dengan total harga sebesar 357 dolar.

Apabila Anda penasaran dengan produk yang memiliki harga termahal dan terendah, silakan jalankan kode berikut untuk melihatnya.

In [ ]:
products_df.sort_values(by="price", ascending=False)

product_id product_type product_name size  colour  price  quantity  \
671          665       Jacket        Parka   XS     red    119        65   
672          666       Jacket        Parka    S     red    119        66   
673          667       Jacket        Parka    M     red    119        56   
674          668       Jacket        Parka    L     red    119        73   
675          669       Jacket        Parka   XL     red    119        64   
...          ...          ...          ...  ...     ...    ...       ...   
1222        1216     Trousers      Pleated    S  indigo     90        41   
1229        1223     Trousers      Pleated    L  violet     90        48   
1228        1222     Trousers      Pleated    M  violet     90        45   
1227        1221     Trousers      Pleated    S  violet     90        50   
1226        1220     Trousers      Pleated   XS  violet     90        58   

                                        description  
671          A red coloured, XS sized, Parka Jacket  
672           A red coloured, S sized, Parka Jacket  
673           A red coloured, M sized, Parka Jacket  
674           A red coloured, L sized, Parka Jacket  
675          A red coloured, XL sized, Parka Jacket  
...                                             ...  
1222   A indigo coloured, S sized, Pleated Trousers  
1229   A violet coloured, L sized, Pleated Trousers  
1228   A violet coloured, M sized, Pleated Trousers  
1227   A violet coloured, S sized, Pleated Trousers  
1226  A violet coloured, XS sized, Pleated Trousers  

[1260 rows x 8 columns]

Selanjutnya, kita bisa menggunakan pivot table untuk mencari informasi terkait produk berdasarkan tipe dan nama produknya. Berikut contoh kode yang bisa Anda gunakan.

In [ ]:

products_df.groupby(by="product_type").agg({
    "product_id": "nunique",
    "quantity": "sum",
    "price":  ["min", "max"]
})


product_id quantity price     
                nunique      sum   min  max
product_type                               
Jacket              420    25387    90  119
Shirt               420    25263    92  119
Trousers            420    25139    90  119

In [ ]:
products_df.groupby(by="product_name").agg({
    "product_id": "nunique",
    "quantity": "sum",
    "price": ["min", "max"]
})

product_id quantity price     
                     nunique      sum   min  max
product_name                                    
Bomber                    35     2083    90   90
Camp Collared             35     2071   112  112
Cardigan                  35     2032   118  118
Cargo Pants               35     2146   106  106
Casual Slim Fit           35     2086   113  113
Chambray                  35     2020   105  105
Chinos                    35     2101   100  100
Coach                     35     2158   115  115
Cords                     35     2260   113  113
Cropped                   35     2085    99   99
Cuban Collar              35     2017    93   93
Denim                     70     4204    92  115
Drawstring                35     2042   104  104
Dress                     35     2125   119  119
Flannel                   35     2135    96   96
Henley                    35     2051    92   92
High-Waisted              35     2198    98   98
Joggers                   35     2107    94   94
Leather                   35     2276   113  113
Linen                     35     2138   116  116
Mandarin Collar           35     2203   108  108
Oxford Cloth              35     2071   114  114
Parka                     35     2201   119  119
Peacoat                   35     2067   102  102
Pleated                   35     2030    90   90
Polo                      35     2196   117  117
Puffer                    35     1946   110  110
Pullover                  35     2184   114  114
Relaxed Leg               35     2002    95   95
Shearling                 35     2169    95   95
Slim-Fit                  35     2155   119  119
Tracksuit Bottoms         35     2038    91   91
Trench Coat               35     2132   112  112
Windbreaker               35     2085   109  109
Wool                      35     1975   111  111

Pivot table di atas dapat memberikan kita gambaran terkait produk yang dijual oleh Fashion Collection.

Sebagai calon praktisi data yang Andal, tentunya Anda penasaran dengan produk yang paling laris. Nah, untuk menjawab pertanyaan ini, kita perlu menyatukan (merge) tabel product_df dan sales_df dengan kode berikut.

In [ ]:
sales_products_df = pd.merge(
    left=sales_df,
    right=products_df,
    how="left",
    left_on="product_id",
    right_on="product_id"
)
sales_products_df.head()

sales_id  order_id  product_id  price_per_unit  quantity_x  total_price  \
0         0         1         218             106           2          212   
1         1         1         481             118           1          118   
2         2         1           2              96           3          288   
3         3         1        1002             106           2          212   
4         4         1         691             113           3          339   

  product_type  product_name size  colour  price  quantity_y  \
0        Shirt      Chambray    L  orange    105          44   
1       Jacket        Puffer    S  indigo    110          62   
2        Shirt  Oxford Cloth    M     red    114          54   
3     Trousers          Wool    M    blue    111          52   
4       Jacket         Parka    S  indigo    119          53   

                                   description  
0   A orange coloured, L sized, Chambray Shirt  
1    A indigo coloured, S sized, Puffer Jacket  
2  A red coloured, M sized, Oxford Cloth Shirt  
3      A blue coloured, M sized, Wool Trousers  
4     A indigo coloured, S sized, Parka Jacket

Pada dasarnya kode di atas akan melakukan proses merge terhadap data product_df dan sales_df. Proses merge tersebut dilakukan menggunakan metode “left”. Berikut hasil dari proses merge tersebut.

ika diperhatikan kembali, hasil dari proses merge di atas memiliki perbedaan antara nilai price_per_unit dan price. Hal ini bisa terjadi karena adanya potongan harga, biaya operasional, dan biaya lainnya.

sekarang kita coba melihat informasi penjualan produk berdasarkan tipenya. dengan membuat pivot table berdasarkan produk type dengan kode berikut

In [ ]:
sales_products_df.groupby(by="product_type").agg({
    "sales_id": "nunique",
    "quantity_x": "sum",
    "total_price": "sum"
})

sales_id  quantity_x  total_price
product_type                                   
Jacket            1676        3343       357026
Shirt             1641        3259       333600
Trousers          1683        3360       341174

Jika Anda perhatikan pivot table di atas, Trousers merupakan tipe produk yang paling laris. Akan tetapi, jika dilihat berdasarkan revenue yang diterima, Jacket merupakan tipe produk yang paling banyak menyumbang revenue perusahaan.

membuat pivot table yang sama untuk melihat informasi penjualan berdasarkan nama produk menggunakan kode di bawah in

In [ ]:

sales_products_df.groupby(by="product_name").agg({
    "sales_id": "nunique",
    "quantity_x": "sum",
    "total_price": "sum"
}).sort_values(by="total_price", ascending=False)

sales_id  quantity_x  total_price
product_name                                        
Denim                   273         527        52399
Casual Slim Fit         154         306        36414
Trench Coat             146         299        35581
Shearling               150         302        35334
Puffer                  140         298        35164
Flannel                 141         281        33158
Cropped                 135         284        32660
Pleated                 147         308        32340
Joggers                 164         334        31062
Chambray                141         290        30740
Parka                   134         269        30397
Bomber                  150         297        29700
Chinos                  135         260        29380
Windbreaker             143         287        29274
Cardigan                134         260        29120
Peacoat                 132         266        28994
Tracksuit Bottoms       130         257        28784
Slim-Fit                154         288        27936
Oxford Cloth            146         289        27744
Drawstring              147         288        27648
High-Waisted            140         278        27522
Camp Collared           139         279        27342
Cargo Pants             134         267        27234
Relaxed Leg             146         301        27090
Cuban Collar            122         254        26670
Henley                  146         281        26414
Wool                    125         249        26394
Leather                 136         263        25774
Mandarin Collar         124         236        25724
Dress                   127         243        25272
Pullover                143         276        25116
Polo                    131         264        25080
Coach                   134         265        24645
Linen                   131         270        24570
Cords                   126         246        23124

Berdasarkan pivot table tersebut, diketahui bahwa produk Denim merupakan produk yang paling laris dan juga paling banyak menyumbang revenue terhadap perusahaan.

Eksplorasi Data all_df

kita perlu membuat sebuah DataFrame baru bernama all_df untuk menampung semua informasi dari keempat tabel yang kita miliki. Berikut contoh kode yang dapat kita gunakan untuk menjalankan proses penggabungan ini.

In [ ]:
all_df = pd.merge(
    left=sales_products_df,
    right=orders_customers_df,
    how="left",
    left_on="order_id",
    right_on="order_id"
)
all_df.head()

sales_id  order_id  product_id  price_per_unit  quantity_x  total_price  \
0         0         1         218             106           2          212   
1         1         1         481             118           1          118   
2         2         1           2              96           3          288   
3         3         1        1002             106           2          212   
4         4         1         691             113           3          339   

  product_type  product_name size  colour  ...     customer_name  gender age  \
0        Shirt      Chambray    L  orange  ...  Annabella Devote     NaN  75   
1       Jacket        Puffer    S  indigo  ...  Annabella Devote     NaN  75   
2        Shirt  Oxford Cloth    M     red  ...  Annabella Devote     NaN  75   
3     Trousers          Wool    M    blue  ...  Annabella Devote     NaN  75   
4       Jacket         Parka    S  indigo  ...  Annabella Devote     NaN  75   

                home_address  zip_code            city            state  \
0  4927 Alice MeadowApt. 960      7787  Sanfordborough  South Australia   
1  4927 Alice MeadowApt. 960      7787  Sanfordborough  South Australia   
2  4927 Alice MeadowApt. 960      7787  Sanfordborough  South Australia   
3  4927 Alice MeadowApt. 960      7787  Sanfordborough  South Australia   
4  4927 Alice MeadowApt. 960      7787  Sanfordborough  South Australia   

     country  status  age_group  
0  Australia  Active    Seniors  
1  Australia  Active    Seniors  
2  Australia  Active    Seniors  
3  Australia  Active    Seniors  
4  Australia  Active    Seniors  

[5 rows x 28 columns]

Proses merge di atas akan menghasilkan DataFrame seperti hasil di atas

Selanjutnya, mari kita coba lihat preferensi pembelian berdasarkan state pelanggan dan tipe produk menggunakan kode di bawah ini.

In [ ]:
all_df.groupby(by=["state", "product_type"]).agg({
    "quantity_x": "sum",
    "total_price": "sum"
})

quantity_x  total_price
state                        product_type                         
Australian Capital Territory Jacket               406        43204
                             Shirt                396        40448
                             Trousers             454        46790
New South Wales              Jacket               451        47998
                             Shirt                431        43980
                             Trousers             392        39766
Northern Territory           Jacket               365        38991
                             Shirt                336        33865
                             Trousers             384        38998
Queensland                   Jacket               499        53511
                             Shirt                417        42506
                             Trousers             453        46045
South Australia              Jacket               461        49090
                             Shirt                509        52685
                             Trousers             455        46041
Tasmania                     Jacket               412        44370
                             Shirt                387        39668
                             Trousers             379        38057
Victoria                     Jacket               359        38203
                             Shirt                390        40206
                             Trousers             414        41948
Western Australia            Jacket               390        41659
                             Shirt                393        40242
                             Trousers             429        43529

Pivot table di atas memberikan kita gambaran terkait tipe produk yang disukai pengguna berdasarkan lokasi negara bagiannya. Jika diperhatikan, tipe produk Jacket (garis merah) paling banyak terjual pada negara bagian Queensland, South Australia, dan New South Wales. Untuk tipe produk lain, Anda bisa melihatnya pada gambar pivot table di atas.

Sekarang Anda bisa melakukan hal yang sama untuk mengetahui selera tipe produk pelanggan berdasarkan gender dan kelompok usia. Berikut contoh kode yang bisa Anda gunakan.

In [ ]:
all_df.groupby(by=["gender", "product_type"]).agg({
    "quantity_x": "sum",
    "total_price": "sum"
})


quantity_x  total_price
gender product_type                         
1.0    Jacket               480        50978
       Shirt                410        41939
       Trousers             453        46071
2.0    Jacket               481        50963
       Shirt                445        45146
       Trousers             481        49295

In [ ]:
all_df.groupby(by=["age_group", "product_type"]).agg({
    "quantity_x": "sum",
    "total_price": "sum"
})

quantity_x  total_price
age_group product_type                         
Adults    Jacket              2331       249262
          Shirt               2254       230723
          Trousers            2342       238256
Seniors   Jacket               738        78752
          Shirt                704        72035
          Trousers             696        70486
Youth     Jacket               274        29012
          Shirt                301        30842
          Trousers             322        32432

pivot table di atas dapat memberikan kita gambaran terkait selera tipe produk pelanggan berdasarkan gender dan kelompok usia. Semua informasi ini akan sangat membantu Anda dalam menjawab pertanyaan terkait preferensi pengguna berdasarkan demografinya.

Tugas

Jawablah beberapa pertanyaan berikut ini berdasarkan EDA yang sudah di lakukan

1.Bagaimana performa penjualan dan revenue perusahaan dalam beberapa bulan terakhir?

2.Produk apa yang paling banyak dan paling sedikit terjual?

3.Bagaimana demografi pelanggan yang kita miliki?

4.Kapan terakhir pelanggan melakukan transaksi?

5.Seberapa sering seorang pelanggan melakukan pembelian dalam beberapa bulan terakhir?

6.Berapa banyak uang yang dihabiskan pelanggan dalam beberapa bulan terakhir?.